### PR TITLE
Issue #34: Create Events Activity

### PR DESCRIPTION
**Pull request by SalosV**
**Create EventActivity** Issue #34 

- Add Layouts 
- Add Adapter
- MVVM implementation

**Testing Done:**

On Emulator: Huawei P20 Lite

**How to test:**
 [ ] Please 
- [ ] Run the application and click on Event Category 
- [ ] You can see the loading while the comics have been loaded 
- [ ] Then you can see the comics in a scrollable list

**How it looks:**
WhatsApp Image 2020-01-01 at 9 02 15 PM
Signed-off-by:

Santiago Lozada <santiagolozada95@gmail.com>

### CODE REVIEW

**Comment by aliceresponde** [comment]
@SalosV please resolve conflics

File: data/src/main/java/com/architect/coders/mu8/data/characters/CharactersResponse.kt [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (jamacado) [inline data/src/main/java/com/architect/coders/mu8/data/characters/CharactersResponse.kt]: is used somewhere because it's empty?

File: presentation/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -14,10 +14,11 @@
         android:supportsRtl="true"
         android:theme="@style/AppTheme"
         tools:ignore="AllowBackup,GoogleAppIndexingWarning">
-
+        <activity android:name=".events.EventsActivity"></activity>
         <activity android:name=".splash.SplashActivity">
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />
+
```
Comment (jamacado) [inline presentation/src/main/AndroidManifest.xml]: Why do I add the space?

File: data/src/main/java/com/architect/coders/mu8/data/events/EventsRepositoryImpl.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package com.architect.coders.mu8.data.events
+
+import com.architect.coders.mu8.data.service.MarvelServiceManager
+import com.architect.coders.mu8.data.utils.MARVEL_PUBLIC_KEY
+import com.architect.coders.mu8.data.utils.TIME_STAMP
+import com.architect.codes.mu8.common.Scope
+import com.architect.codes.mu8.events.Event
+import com.architect.codes.mu8.events.EventsRepository
+
+class EventsRepositoryImpl(private val mapper: EventsMapper) : EventsRepository, Scope by Scope.Implementation() {
```
Comment (recaldev) [inline data/src/main/java/com/architect/coders/mu8/data/events/EventsRepositoryImpl.kt]: Repository should not handle the scope.

File: data/src/main/java/com/architect/coders/mu8/data/events/EventsMapper.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package com.architect.coders.mu8.data.events
+
+import com.architect.coders.mu8.data.mapper.BaseResponseMapper
+import com.architect.coders.mu8.data.utils.replaceHttps
+import com.architect.codes.mu8.events.Event
+
+class EventsMapper : BaseResponseMapper<EventsResponse, Event> () {
+    override fun transform(input: EventsResponse): Event {
```
Comment (recaldev) [inline data/src/main/java/com/architect/coders/mu8/data/events/EventsMapper.kt:9]: Put a new line above this to improve reading.

File: data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt [hunk 5]
```diff
@@ -14,7 +15,7 @@ private const val OFFSET = "offset"
 private const val LIMIT = "limit"
 
 /**
- * All marvel API Calls requires {@Link [TIME_STAMP]}, {@Link[API_KEY]}, and {@Link [HASH]}
+ * All marvel API Calls requires {@Link [TIME_STAMP]}, {@Link[API_KEY]}, and {@Link [HASH] }
```
Comment (recaldev) [inline data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt]: Revert this change.

File: domain/src/main/java/com/architect/codes/mu8/events/Event.kt [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package com.architect.codes.mu8.events
+
+data class Event(
```
Comment (recaldev) [inline domain/src/main/java/com/architect/codes/mu8/events/Event.kt:5]: I think this data class fits in one line.

File: domain/src/main/java/com/architect/codes/mu8/events/EventsUseCase.kt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package com.architect.codes.mu8.events
+
+interface EventsUseCase {
+    suspend fun getEvents(): List<Event>
+}
+
+class EventsUserCaseImpl(private val eventsRepository: EventsRepository) : EventsUseCase {
+    override suspend fun getEvents(): List<Event> = eventsRepository()
```
Comment (recaldev) [inline domain/src/main/java/com/architect/codes/mu8/events/EventsUseCase.kt]: You can use here `operator invoke()` too since there is only one method to call.

File: presentation/build.gradle [hunk 8]
```diff
@@ -44,9 +44,11 @@ dependencies {
     implementation SupportDependencies.firebaseAnalytics
 
     implementation LibsDependencies.glide
+    implementation 'androidx.appcompat:appcompat:1.1.0'
+    implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
```
Comment (recaldev) [inline presentation/build.gradle]: Why did you add this two dependencies? They are already added.

File: presentation/src/main/AndroidManifest.xml [hunk 9]
```diff
@@ -14,10 +14,11 @@
         android:supportsRtl="true"
         android:theme="@style/AppTheme"
         tools:ignore="AllowBackup,GoogleAppIndexingWarning">
-
+        <activity android:name=".events.EventsActivity"></activity>
```
Comment (recaldev) [inline presentation/src/main/AndroidManifest.xml]: Put this below with the other activities and close the activity tag at the end of the name like this `<activity android:name=".events.EventsActivity"/>`

File: presentation/src/main/AndroidManifest.xml (hunk 2 above)
Comment (recaldev) [inline presentation/src/main/AndroidManifest.xml]: Remove space.

File: presentation/src/main/java/com/architect/coders/mu8/events/EventsActivity.kt [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        recyclerView.layoutManager = LinearLayoutManager(this)
+        recyclerView.adapter = adapter
+
+        viewModel.model.observe(this, Observer(::updateUi))
+    }
+
+    private fun updateUi(model: EventsUiModel) {
+        progress.visibility = if (model == Loading) VISIBLE else GONE
+
+        when (model) {
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/events/EventsActivity.kt:51]: You can replace this by a single if since you're not using more states from the model, or you can  add the progress here. Also, you can add the navigation event instead and put a toast for now.

File: presentation/src/main/java/com/architect/coders/mu8/events/EventsAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        holder.bind(events[position])
+    }
+
+    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+
+        private var image: ImageView = itemView.findViewById(R.id.image)
+        private var title: TextView = itemView.findViewById(R.id.title)
+
+        fun bind(item: Event) {
+
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/events/EventsAdapter.kt]: Remove extra space.

File: presentation/src/main/java/com/architect/coders/mu8/events/EventsViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+import androidx.lifecycle.MutableLiveData
+import com.architect.coders.mu8.utils.ScopedViewModel
+import com.architect.codes.mu8.events.EventsUseCase
+import kotlinx.coroutines.launch
+
+class EventsViewModel(private val eventsUseCase: EventsUseCase) : ScopedViewModel() {
+
+    private val _model = MutableLiveData<EventsUiModel>()
+    val model: LiveData<EventsUiModel>
+    get() { if (_model.value == null) getEvents()
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/events/EventsViewModel.kt]: This is missing the right format. Please fix it to improve its reading.

File: presentation/src/main/res/layout/activity_events.xml [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+    <ProgressBar
+        android:id="@+id/progress"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/include_toolbar" />
+
```
Comment (recaldev) [inline presentation/src/main/res/layout/activity_events.xml]: Remove extra space.

File: presentation/src/main/res/layout/event_item.xml [hunk 14]
```diff
@@ -0,0 +1,30 @@
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content">
+
+    <ImageView
+        android:id="@+id/image"
+        android:layout_width="match_parent"
+        android:layout_height="200dp"
```
Comment (recaldev) [inline presentation/src/main/res/layout/event_item.xml]: Use dimens for this value.

File: presentation/src/main/res/layout/event_item.xml [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+        android:contentDescription="@null"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <TextView
+        android:id="@+id/title"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginEnd="10dp"
```
Comment (recaldev) [inline presentation/src/main/res/layout/event_item.xml]: Use dimens for this value.

File: presentation/src/main/res/layout/event_item.xml [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+        android:layout_marginBottom="10dp"
```
Comment (recaldev) [inline presentation/src/main/res/layout/event_item.xml]: Use dimens for this value.

File: presentation/src/main/res/layout/event_item.xml [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+        android:elevation="5dp"
```
Comment (recaldev) [inline presentation/src/main/res/layout/event_item.xml]: Use dimens for this value.

File: presentation/src/main/res/layout/event_item.xml [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+        android:shadowColor="@android:color/black"
+        android:shadowRadius="50"
```
Comment (recaldev) [inline presentation/src/main/res/layout/event_item.xml]: Use dimens for this value.

**Review by recaldev — CHANGES_REQUESTED** [review]
Please review the comments and make the changes. Also, wouldn't this going to be a PR with a map and the events on it? Why did you make this instead?

File: domain/src/main/java/com/architect/codes/mu8/events/Event.kt (hunk 6 above)
Comment (SalosV) [reply domain/src/main/java/com/architect/codes/mu8/events/Event.kt:5]: Done

File: presentation/build.gradle (hunk 8 above)
Comment (SalosV) [reply presentation/build.gradle]: DoneI deleted them, they were not necessary

File: presentation/src/main/AndroidManifest.xml (hunk 9 above)
Comment (SalosV) [reply presentation/src/main/AndroidManifest.xml]: Done

File: presentation/src/main/AndroidManifest.xml (hunk 2 above)
Comment (SalosV) [reply presentation/src/main/AndroidManifest.xml]: Done

File: domain/src/main/java/com/architect/codes/mu8/events/EventsUseCase.kt (hunk 7 above)
Comment (SalosV) [reply domain/src/main/java/com/architect/codes/mu8/events/EventsUseCase.kt]: Done

File: data/src/main/java/com/architect/coders/mu8/data/events/EventsMapper.kt (hunk 4 above)
Comment (SalosV) [reply data/src/main/java/com/architect/coders/mu8/data/events/EventsMapper.kt:9]: Done

File: presentation/src/main/java/com/architect/coders/mu8/events/EventsAdapter.kt (hunk 11 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/events/EventsAdapter.kt]: Done

File: presentation/src/main/res/layout/activity_events.xml (hunk 13 above)
Comment (SalosV) [reply presentation/src/main/res/layout/activity_events.xml]: Done

File: presentation/src/main/res/layout/event_item.xml (hunk 14 above)
Comment (SalosV) [reply presentation/src/main/res/layout/event_item.xml]: Done

File: presentation/src/main/java/com/architect/coders/mu8/events/EventsViewModel.kt (hunk 12 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/events/EventsViewModel.kt]: Done

File: presentation/src/main/res/layout/event_item.xml (hunk 15 above)
Comment (SalosV) [reply presentation/src/main/res/layout/event_item.xml]: Done

File: presentation/src/main/res/layout/event_item.xml (hunk 16 above)
Comment (SalosV) [reply presentation/src/main/res/layout/event_item.xml]: Done

File: presentation/src/main/res/layout/event_item.xml (hunk 17 above)
Comment (SalosV) [reply presentation/src/main/res/layout/event_item.xml]: Done

File: presentation/src/main/res/layout/event_item.xml (hunk 18 above)
Comment (SalosV) [reply presentation/src/main/res/layout/event_item.xml]: Done

File: data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt (hunk 5 above)
Comment (SalosV) [reply data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt]: Done

File: presentation/src/main/java/com/architect/coders/mu8/events/EventsActivity.kt (hunk 10 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/events/EventsActivity.kt:51]: Done

File: data/src/main/java/com/architect/coders/mu8/data/events/EventsRepositoryImpl.kt (hunk 3 above)
Comment (SalosV) [reply data/src/main/java/com/architect/coders/mu8/data/events/EventsRepositoryImpl.kt]: Done

File: data/src/main/java/com/architect/coders/mu8/data/events/EventsResponse.kt [hunk 19]
```diff
@@ -0,0 +1,11 @@
+package com.architect.coders.mu8.data.events
+
+import com.architect.coders.mu8.data.response.common.ThumbnailResponse
+import com.architect.codes.mu8.utils.EMPTY_STRING
+
+data class EventsResponse(
+    val id: Long = 0L,
+    val title: String = EMPTY_STRING,
+    val description: String = EMPTY_STRING,
+    val thumbnail: ThumbnailResponse
```
Comment (recaldev) [inline data/src/main/java/com/architect/coders/mu8/data/events/EventsResponse.kt]: put the default value for this as en empty object, `ThumbnailResponse` already have default values to avoid nullability.

File: data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt [hunk 20]
```diff
@@ -35,4 +36,11 @@ interface MarvelServiceAPI {
...
         @Query(LIMIT) limit: Int
     ): Response<BaseResponse<CharactersResponse>>
-}
\ No newline at end of file
+
+    @GET("/v1/public/events")
+    suspend fun getAllEvents(
+        @Query(TIME_STAMP) timeStamp: String,
+        @Query(API_KEY) apiKey: String,
+        @Query(HASH) hash: String
```
Comment (recaldev) [inline data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt]: Please also use the offset and limit properties to get a good number of events and not the default one.

File: domain/src/main/java/com/architect/codes/mu8/events/Event.kt [hunk 21]
```diff
@@ -0,0 +1,3 @@
+package com.architect.codes.mu8.events
+
+data class Event(val id: Long, val title: String, val thumbnail: String)
```
Comment (recaldev) [inline domain/src/main/java/com/architect/codes/mu8/events/Event.kt]: Can we set some default values?

**Review by recaldev — CHANGES_REQUESTED** [review]

File: data/src/main/java/com/architect/coders/mu8/data/events/EventsResponse.kt (hunk 19 above)
Comment (SalosV) [reply data/src/main/java/com/architect/coders/mu8/data/events/EventsResponse.kt]: Done

File: domain/src/main/java/com/architect/codes/mu8/events/Event.kt (hunk 21 above)
Comment (SalosV) [reply domain/src/main/java/com/architect/codes/mu8/events/Event.kt]: Done

File: data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt (hunk 20 above)
Comment (SalosV) [reply data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt]: Done

**Review by recaldev — APPROVED** [review]
Looking good for me.

**Review by jamacado — APPROVED** [review]
